legion/engine/llri-vk/source.cpp

```diff
@@ -1,14 +1,259 @@
+
+#include <vector>
+#include <map>
+#include <vulkan/vulkan.hpp>
 
 namespace legion::graphics::llri
 {
-    namespace detail
+    result createInstance(const instance_desc& desc, Instance* instance)
     {
-        static constexpr unsigned VKStaticId = 42;
+        if (instance == nullptr)
+            return result::ErrorInvalidUsage;
+        if (desc.numExtensions > 0 && desc.extensions == nullptr)
+            return result::ErrorInvalidUsage;
+
+        auto* result = new InstanceT();
+
+        std::vector<const char*> layers;
+        std::vector<const char*> extensions;
+
+        auto availableLayers = internal::queryAvailableLayers();
+        auto availableExtensions = internal::queryAvailableExtensions();
+
+        void* pNext = nullptr;
+
+        //Variables that need to be stored outside of scope
+        vk::ValidationFeaturesEXT features; 
+        std::vector<vk::ValidationFeatureEnableEXT> enables;
+
+        for (uint32_t i = 0; i < desc.numExtensions; i++)
+        {
+            auto& extension = desc.extensions[i];
+            switch(extension.type)
+            {
+                case instance_extension_type::APIValidation:
+                {
+                    if (extension.apiValidation.enable)
+                        layers.push_back("VK_LAYER_KHRONOS_validation");
+                    break;
+                }
+                case instance_extension_type::GPUValidation:
+                {
+                    if (extension.gpuValidation.enable)
+                    {
+                        enables = {
+                            vk::ValidationFeatureEnableEXT::eGpuAssisted,
+                            vk::ValidationFeatureEnableEXT::eGpuAssistedReserveBindingSlot
+                        };
+                        features = vk::ValidationFeaturesEXT(enables, {});
+                        features.pNext = pNext; //Always apply pNext backwards to simplify optional chaining
+                        pNext = &features;
+                    }
+                    break;
+                }
+                default:
+                {
+                    llri::destroyInstance(result);
+                    return result::ErrorExtensionNotSupported;
+                }
+            }
+        }
+
+        vk::ApplicationInfo appInfo{ desc.applicationName, VK_MAKE_VERSION(0, 0, 0), "Legion::LLRI", VK_MAKE_VERSION(0, 0, 1), VK_HEADER_VERSION_COMPLETE };
+        vk::InstanceCreateInfo ci{ {}, &appInfo, (uint32_t)layers.size(), layers.data(), (uint32_t)extensions.size(), extensions.data() };
+        ci.pNext = pNext;
+
+        vk::Instance vulkanInstance = nullptr;
+        const vk::Result createResult = vk::createInstance(&ci, nullptr, &vulkanInstance);
+
+        if (createResult != vk::Result::eSuccess)
+        {
+            llri::destroyInstance(result);
+            return internal::mapVkResult(createResult);
+        }
+
+        result->m_ptr = vulkanInstance;
+        *instance = result;
```

Comment:
another reason why type aliasing a raw pointer is generally discouraged is because rn it looks like you copy result over into the address at instance, but then you don't delete result. It makes sense because instance is actually an `InstanceT**` but the type alias does create an unnecessary confusing situation.

Reply:
Agreed, I'll switch away from aliasing to just letting users hold Instance* pointers directly